Hail fails when aggregating over a lazily filtered aggregable

Stack trace from @lfrancioli ([full trace](https://nealelab.slack.com/files/laurent/F3P268282/error.txt))

```
Caused by: java.lang.AssertionError: assertion failed
    at scala.Predef$.assert(Predef.scala:156)
    at org.broadinstitute.hail.utils.richUtils.RichIterable$$anon$4$$anon$10.next(RichIterable.scala:71)
    at org.broadinstitute.hail.methods.Aggregators$$anonfun$buildVariantAggregations$1.apply(Aggregators.scala:54)
    at org.broadinstitute.hail.methods.Aggregators$$anonfun$buildVariantAggregations$1.apply(Aggregators.scala:45)
    at org.broadinstitute.hail.driver.AnnotateVariantsExpr$$anonfun$4$$anonfun$apply$1.apply(AnnotateVariantsExpr.scala:51)
    ⋮
```

The iterator returned by the genotype stream has an additional constraint (over the `Iterator[T]` interface) that `hasNext` must be called before every call to `next`. The failing assertion verifies that.

--- a/src/main/scala/org/broadinstitute/hail/methods/Aggregators.scala
+++ b/src/main/scala/org/broadinstitute/hail/methods/Aggregators.scala
@@ -50,8 +50,9 @@ object Aggregators {
 
       val gsIt = gs.iterator
       var i = 0
-      while (i < localSamplesBc.value.size) {
-        localA(3) = gsIt.next()
+      // gsIt assume hasNext is always called before next
+      while (gsIt.hasNext) {
+        localA(3) = gsIt.next
         localA(4) = localSamplesBc.value(i)
         localA(5) = localAnnotationsBc.value(i)
 
@@ -115,7 +116,8 @@ object Aggregators {
 
       val gsIt = gs.iterator
       var i = 0
-      while (i < localSamplesBc.value.size) {
+      // gsIt assume hasNext is always called before next
+      while (gsIt.hasNext) {
         localA(1) = localSamplesBc.value(i)
         localA(2) = localSampleAnnotationsBc.value(i)
         localA(3) = gsIt.next

--- a/src/test/scala/org/broadinstitute/hail/methods/AggregatorSuite.scala
+++ b/src/test/scala/org/broadinstitute/hail/methods/AggregatorSuite.scala
@@ -344,4 +344,18 @@ class AggregatorSuite extends SparkSuite {
     }
     p.check()
   }
+
+  @Test def filterMap() {
+    val vds = VariantSampleMatrix.gen(sc, VSMSubgen.random.copy(sampleIdGen=Gen.const(Array("a", "b")))).sample()
+    var s = State(sc, sqlContext, vds)
+    s = FilterSamplesExpr.run(s, Array("-c", "s == \"a\"", "--keep"))
+    s = AnnotateVariantsExpr.run(s, Array("-c", "va.result = gs.map(id => 1).sum()"))
+
+    val (_, result) = s.vds.queryVA("va.result")
+
+    val va = s.vds.variantsAndAnnotations.map(_._2).collect().head
+    result(va).foreach { result =>
+      assert(result == 1)
+    }
+  }
 }

